source/distributed/tria.cc

```diff
@@ -3272,6 +3470,11 @@ namespace parallel
           // fix all the flags to make sure we have a consistent mesh
           this->prepare_coarsening_and_refinement ();
 
+          // enforce 2:1 mesh balance over periodic boundaries
+          if (this->smooth_grid & dealii::Triangulation<dim,spacedim>::limit_level_difference_at_vertices)
+            enforce_mesh_balance_over_periodic_boundaries(*this,
```

Comment:
Regarding the single iteration: It can require more iterations, but there is already an iteration `while(mesh_changed)` around that code block in the distributed triangulation.

Regarding `prepare_coarsening_and_refinement`: I have rewritten the patch that way. The overridden `prepare_coarsening_and_refinement` method calls the implementation of the base class and `enforce_mesh_balance_over_periodic_boundaries`, looping until nothing changes any more. Without periodic boundaries (or no `limit_level_difference_at_vertices` which is only set when the multigrid hierarchy is requested), the method is called once. Otherwise, we will ensure that the two methods have 'converged'.